core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java

```diff
@@ -682,4 +780,462 @@ private boolean hasWriteBehindSharedStore(PersistenceConfiguration persistenceCo
+
+                  InnerPublisherSubscription<K, I, R>[] publisherArray = new InnerPublisherSubscription[concurrentPublishers];
+                  for (int i = 0; i < concurrentPublishers - 1; ++i) {
+                     publisherArray[i] = new InnerPublisherSubscription<>(this, targetBatchSize, targetSupplier,
+                           excludedKeys, currentTopology);
+                  }
+                  // Submit the local target last if necessary (otherwise is a normal submission)
+                  // This is done last as we want to send all the remote requests first and only process the local
+                  // container concurrently with the remote requests
+                  if (localSegments != null) {
+                     publisherArray[concurrentPublishers - 1] = new InnerPublisherSubscription<>(this, targetBatchSize,
+                           targetSupplier, excludedKeys, currentTopology, new AbstractMap.SimpleEntry<>(localAddress, localSegments));
+                  } else {
+                     publisherArray[concurrentPublishers - 1] = new InnerPublisherSubscription<>(this, targetBatchSize,
+                           targetSupplier, excludedKeys, currentTopology);
+                  }
+
+                  return Flowable.mergeArray(publisherArray);
+               }, 4)
+               .repeatUntil(() -> {
+                  boolean complete = segmentsToComplete.isEmpty();
+                  if (trace) {
+                     if (complete) {
+                        log.tracef("All segments complete for %s", requestId);
+                     } else {
+                        log.tracef("Segments %s not completed - retrying", segmentsToComplete);
+                     }
+                  }
+                  return complete;
+               });
+
+         if (completedSegmentConsumer != null) {
+            ByRef<R> previousValue = new ByRef<>(null);
+            valuesFlowable = valuesFlowable.doOnNext(value -> {
+               R previous = previousValue.get();
+               if (previous != null) {
+                  IntSet segments = enqueuedSegmentNotifiers.remove(previous);
+                  if (segments != null) {
+                     if (trace) {
+                        log.tracef("Enqueued value %s has been returned, completing segments %s",
+                              Util.toStr(previous), segments);
+                     }
+                     segments.forEach(completedSegmentConsumer);
+                  }
+               }
+               previousValue.set(value);
+            }).doOnComplete(() -> enqueuedSegmentNotifiers.forEach(
+                  (k, segments) -> {
+                     if (trace) {
+                        log.tracef("Notifying of completed segments %s due to publisher is complete", segments);
+                     }
+                     segments.forEach(completedSegmentConsumer);
+                  })
+            );
```

Comment:
Should we also handle `doOnError`?

Reply:
What were you thinking would go in there? We don't want to complete the segments. The error would be propagated to the caller, so the entire Subscription would be cancelled anyways.